File: src/main/java/org/oewntk/model/LexGroupings.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.*;

import static java.util.stream.Collectors.*;

public class LexGroupings
{
	/**
	 * Group lexes by CS lemma
	 *
	 * @param lexes lexes
	 * @return lexes grouped by CS lemma
	 */
	public static Map<String, Collection<Lex>> lexesByLemma(final Collection<Lex> lexes)
	{
		return Groupings.groupBy(lexes, Lex::getLemma);
	}

	/**
	 * Group lexes by LC lemma
	 *
	 * @param lexes lexes
	 * @return lexes grouped by LCS lemma
	 */
	public static Map<String, Collection<Lex>> lexesByLCLemma(final Collection<Lex> lexes)
	{
		return Groupings.groupBy(lexes, Lex::getLCLemma);
	}

	/**
	 * Hypermap (LCLemma -> CSLemma -> lexes)
	 *
	 * @param model model
	 * @return 2-tier hypermap (LCLemma -> CSLemma -> lexes)
	 */
	public static Map<String, Map<String, Collection<Lex>>> hyperMapByLCLemmaByLemma(final CoreModel model)
	{
		return model.getLexesByLemma().entrySet().stream() //
				.collect(groupingBy(e -> e.getKey().toLowerCase(Locale.ENGLISH), toMap(Map.Entry::getKey, Map.Entry::getValue)));
	}

	/**
	 * CSLemmas grouped by LCLemma
	 * baroque -> (Baroque, baroque)
	 *
	 * @param model model
	 * @return CS lemmas by LC lemmas
	 */
	public static Map<String, List<String>> cSLemmasByLCLemma(final CoreModel model)
	{
		return model.lexes.stream() //
				.map(Lex::getLemma) //
				.collect(groupingBy(l -> l.toLowerCase(Locale.ENGLISH), TreeMap::new, toList()));
	}

	/**
	 * CSLemmas for LCLemma
	 *
	 * @param model model
	 * @return CS lemmas for given LC lemma
	 */
	public static List<String> cSLemmasForLCLemma(final CoreModel model, final String lcLemma)
	{
		return cSLemmasByLCLemma(model).get(lcLemma);
	}

	// counts

	/**
	 * Counts of CS lemmas by LC lemma
	 *
	 * @param model model
	 * @return counts of CS lemmas by LC lemmas
	 */
	public static Map<String, Long> countsByLCLemma(final CoreModel model)
	{
		return Groupings.countsBy(model.lexes.stream().map(Lex::getLemma), lemma -> lemma.toLowerCase(Locale.ENGLISH));
	}

	/**
	 * Counts of CS lemmas by LC lemma, sme as above but retain entries that have count > 1
	 *
	 * @param model model
	 * @return counts of CS lemmas by LC lemmas, with count > 2
	 */
	public static Map<String, Long> multipleCountsByICLemma(final CoreModel model)
	{
		return Groupings.multipleCountsBy(model.lexes.stream().map(Lex::getLemma), lemma -> lemma.toLowerCase(Locale.ENGLISH));
	}

	/**
	 * CS lemmas by LC lemmas, retain entries that have count > 1
	 *
	 * @param model model
	 * @return CS lemmas by LC lemmas, with count > 2
	 */
	public static Map<String, List<String>> cSLemmasByLCLemmaHavingMultipleCount(final CoreModel model)
	{
		return Groupings.groupByHavingMultipleCount(model.lexes.stream().map(Lex::getLemma), lemma -> lemma.toLowerCase(Locale.ENGLISH));
	}
}

File: src/main/java/org/oewntk/model/LexGroupings.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import java.util.*

object LexGroupings {

    /**
     * Group lexes by case-sensitive lemma
     *
     * @param lexes lexes
     * @return lexes grouped by CS lemma
     */
    fun lexesByLemma(lexes: Collection<Lex>): Map<String, Collection<Lex>> {
        return lexes
            .groupBy(Lex::lemma)
            .mapValues { it.value.toSet() }
    }

    /**
     * Group lexes by lower-cased lemma
     *
     * @param lexes lexes
     * @return lexes grouped by LCS lemma
     */
    fun lexesByLCLemma(lexes: Collection<Lex>): Map<String, Collection<Lex>> {
        return lexes
            .groupBy { it.lCLemma }
            .mapValues { it.value.toSet() }
    }

    /**
     * CSLemmas grouped by LCLemma
     * ```
     * baroque -&gt; (Baroque, baroque)
     *```
     * @param model model
     * @return CS lemmas by LC lemmas
     */
    fun cSLemmasByLCLemma(model: CoreModel): Map<String, Set<String>> {
        return model.lexes
            .map(Lex::lemma)
            .groupBy { it.lowercase(Locale.ENGLISH) }
            .mapValues { it.value.toSortedSet() }
    }

    /**
     * CSLemmas for LCLemma
     *
     * @param model   model
     * @param lcLemma lower-cased lemma
     * @return CS lemmas for given LC lemma
     */
    fun cSLemmasForLCLemma(model: CoreModel, lcLemma: String): Set<String> {
        return cSLemmasByLCLemma(model)[lcLemma]!!
    }

    // hyper maps

    /**
     * Hypermap (LCLemma to CSLemma to lexes)
     *
     * @param model model
     * @return 2-tier hypermap
     * ```
     * (LCLemma -> CSLemma -> lexes)
     * ```
     */
    fun hyperMapByLCLemmaByLemma(model: CoreModel): Map<String, Map<String, Collection<Lex>>> {
        return model.lexesByLemma!!.entries // entries: setOf(lemma to lexes)
            .groupBy { entry -> entry.key.lowercase() } // groupBy: mapOf(lclemma to listOf(lemma to lexes)), entry: lemma to lexes
            .mapValues { values -> // values: lcLemma to listOf(lemma to lexes))
                values.value // value: listOf(lemma to lexes)
                    .associateBy { it.key } // mapOf(lemma to lexes)
                    .mapValues { it.value.value } // it: lemma to (lemma to lexes), it.value: lemma to lexes, it.value.value: lexes
            }
    }

    // counts

    /**
     * Counts of CS lemmas by LC lemma
     *
     * @param model model
     * @return counts of CS lemmas by LC lemmas
     */
    fun countsByLCLemma(model: CoreModel): Map<String, Long> {
        return model.lexes
            .map(Lex::lemma)
            .groupBy { it.lowercase(Locale.ENGLISH) }
            .toSortedMap(naturalOrder())
            .mapValues { it.value.toSet().size.toLong() }
    }

    /**
     * Counts of CS lemmas by LC lemma, same as above but retain entries that have count &gt; 2
     *
     * @param model model
     * @return counts of CS lemmas by LC lemmas, with count &gt; 2
     */
    fun multipleCountsByICLemma(model: CoreModel): Map<String, Long> {
        return model.lexes
            .map(Lex::lemma)
            .groupBy { it.lowercase(Locale.ENGLISH) }
            .mapValues { it.value.toSet().size.toLong() }
            .toList()
            .filter { it.second > 1L }
            .toMap()
            .toSortedMap(naturalOrder())
    }

    /**
     * CS lemmas by LC lemmas, retain entries that have count &gt; 2
     *
     * @param model model
     * @return CS lemmas by LC lemmas, with count &gt; 2
     */
    fun cSLemmasByLCLemmaHavingMultipleCount(model: CoreModel): Map<String, Set<String>> {
        return Groupings.groupByHavingMultipleCount(model.lexes.map(Lex::lemma)) { it.lowercase() }
    }
}
